athena-dynamodb/athena-dynamodb.yaml

```diff
@@ -45,35 +52,94 @@ Resources:
-      Policies:
-        - Statement:
-            - Action:
-                - dynamodb:DescribeTable
-                - dynamodb:ListSchemas
-                - dynamodb:ListTables
-                - dynamodb:Query
-                - dynamodb:Scan
-                - glue:GetTableVersions
-                - glue:GetPartitions
-                - glue:GetTables
-                - glue:GetTableVersion
-                - glue:GetDatabases
-                - glue:GetTable
-                - glue:GetPartition
-                - glue:GetDatabase
-                - athena:GetQueryExecution
-                - s3:ListAllMyBuckets
-              Effect: Allow
-              Resource: '*'
-          Version: '2012-10-17'
-        #S3CrudPolicy allows our connector to spill large responses to S3. You can optionally replace this pre-made policy
-        #with one that is more restrictive and can only 'put' but not read,delete, or overwrite files.
-        - S3CrudPolicy:
-            BucketName: !Ref SpillBucket
\ No newline at end of file
+      Role: !GetAtt FunctionRole.Arn
+
+  FunctionRole:
+    Type: AWS::IAM::Role
+    Properties:
+      ManagedPolicyArns:
+        - "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
+      AssumeRolePolicyDocument:
+        Version: 2012-10-17
+        Statement:
+          - Effect: Allow
+            Principal:
+              Service:
+                - lambda.amazonaws.com
+            Action:
+              - "sts:AssumeRole"
+
+  FunctionExecutionPolicy:
+    Type: "AWS::IAM::Policy"
+    Properties:
+      PolicyName: FunctionExecutionPolicy
+      PolicyDocument:
+        Version: 2012-10-17
+        Statement:
+          - Action:
+             - dynamodb:DescribeTable
+             - dynamodb:ListSchemas
+             - dynamodb:ListTables
+             - dynamodb:Query
+             - dynamodb:Scan
+             - glue:GetTableVersions
+             - glue:GetPartitions
+             - glue:GetTables
+             - glue:GetTableVersion
+             - glue:GetDatabases
+             - glue:GetTable
+             - glue:GetPartition
+             - glue:GetDatabase
+             - athena:GetQueryExecution
+             - s3:ListAllMyBuckets
+            Effect: Allow
+            Resource: '*'
+          - Action:
+             - s3:GetObject
+             - s3:ListBucket
+             - s3:GetBucketLocation
+             - s3:GetObjectVersion
+             - s3:PutObject
+             - s3:PutObjectAcl
+             - s3:GetLifecycleConfiguration
+             - s3:PutLifecycleConfiguration
+             - s3:DeleteObject
+            Effect: Allow
+            Resource:
+              - Fn::Sub:
+                - arn:${AWS::Partition}:s3:::${bucketName}
+                - bucketName:
+                    Ref: SpillBucket
+              - Fn::Sub:
+                - arn:${AWS::Partition}:s3:::${bucketName}/*
+                - bucketName:
+                    Ref: SpillBucket
```

Comment:
Why did we change this?

Reply:
This is the S3CrudPolicy but the S3CrudPolicy is an AWS SAM Policy Templates (https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-policy-template-list.html#s3-crud-policy) and can't be referenced in a AWS::IAM::Policy so I basically had to copy paste its content into the policy. I had to use an AWS::IAM::Policy because it is the only way I see to conditionally create it or not based on if KMSKeyId is passed or not. So basically from line 105 to line 124 is equivalent to the previous code:
        - S3CrudPolicy:
            BucketName: !Ref SpillBucket

Reply:
Ack.